imap/jmap_util.c

```diff
@@ -751,14 +744,29 @@ EXPORTED void jmap_decode_to_utf8(const char *charset, int encoding,
                     struct char_counts guess_counts =
                         charset_count_validutf8(guess, strlen(guess));
                     if ((guess_counts.valid > counts.valid) &&
-                        (obj->confidence >= confidence)) {
-                        free(text);
-                        text = guess;
-                        counts = guess_counts;
-                    }
-                    else {
-                        free(guess);
+                            (obj->confidence >= confidence)) {
+                        // libchardet might have guessed a single-byte character encoding,
+                        // which will result in a large number of valid characters and no
+                        // invalid or replacement characters. An indicator that the guess
+                        // was wrong is if the number of printable characters decreased
+                        // in relation to all valid characters, including control chars.
+                        double printable = counts.valid ?
+                            (counts.valid - counts.cntrl) /
+                            ((double) counts.valid) : 0;
```

Comment:
I might write this as something like:

```
double printable = counts.valid
    ? 1.0 * (counts.valid - counts.cntrl) / counts.valid
    : 0;
```

My thinking is:
* the `?:` being left-aligned in a column make them very easy to scan for and see the relationship between the expressions.  I'd prefer to align them with the `=` too, but I don't think there is enough horizontal space here.
* the true-expression being on its own line rather than wrapped helps too
* we're constrained for horizontal space, and `1.0 *` is fewer chars than `(double)`.  The compiler will probably optimise out the multiplication, and even if it doesn't, the multiplication costs nothing compared to the division we're about to do anyway.
* any term in an expression being `double` is enough to promote every other term to `double`, so it doesn't matter whether we explicitly promote the dividend or the divisor (as long as we promote one of them). Sticking the promotion out the front makes it a little clearer up front that this will be a floating point division.  (In either of our implementations, the integer subtraction in the bracketed subexpression won't be promoted to floating point until after its integer result is computed, but it doesn't matter either way.)

This isn't a "you must change it", just a "these are the things I think about when deciding how to wrap a long line". :)

The `guess_printable` computation below might not work quite the same, because I think the true-expression may be too long to fit on a single line at this level of indentation.  In that case I would wrap the true-expression with the second line indented further and starting from `/ ...`.

Reply:
Thanks, the line break just was my editor applying its formatting rules. Also applied your other suggested changes.